tools/cp-caps/rcp_caps_test.py

```diff
@@ -585,44 +572,84 @@ def __test_diag_repeat(self):
-        packets = 100
-        threshold = 80
-        channel = 20
-        cmd_diag_frame = f'diag frame {frame}'
-        commands = [cmd_diag_frame, f'diag send {packets}', f'diag stats', f'diag stats clear']
+    def __test_send_formated_frames_retries(self,
+                                            sender: OTCI,
+                                            receiver: OTCI,
+                                            frame: Frame,
+                                            max_send_retries: int = 5):
+        for i in range(0, max_send_retries):
+            if self.__test_send_formated_frame(sender, receiver, frame):
+                return True
 
-        if self.__support_commands(commands):
-            sender.wait(1)
-            sender.diag_set_channel(channel)
-            receiver.diag_set_channel(channel)
-            receiver.diag_radio_receive()
+        return False
+
+    def __test_send_formated_frame(self, sender: OTCI, receiver: OTCI, frame: Frame):
+        sender.factory_reset()
+        receiver.factory_reset()
+
+        sender.diag_start()
+        receiver.diag_start()
+
+        channel = 11
+        num_sent_frames = 1
+
+        sender.diag_set_channel(channel)
+        receiver.diag_set_channel(channel)
+        receiver.diag_radio_receive()
+        receiver.diag_set_radio_receive_filter_dest_mac_address(frame.dst_address)
+        receiver.diag_enable_radio_receive_filter()
+
+        result_queue = queue.Queue()
+        receive_task = threading.Thread(target=self.__radio_receive_task,
+                                        args=(receiver, num_sent_frames, result_queue),
+                                        daemon=True)
+        receive_task.start()
+
+        sender.wait(0.1)
 
-            sender.diag_stats_clear()
-            sender.diag_stats_clear()
+        sender.diag_frame(frame.tx_frame,
+                          is_security_processed=True,
+                          max_csma_backoffs=4,
+                          max_frame_retries=4,
+                          csma_ca_enabled=True)
+        sender.diag_send(num_sent_frames, is_async=False)
 
-            sender.diag_frame(frame)
-            sender.diag_send(packets, None)
-            sender.wait(1)
-            sender_stats = sender.diag_get_stats()
-            receiver_stats = receiver.diag_get_stats()
+        receive_task.join()
 
-            ret = sender_stats['sent_success_packets'] == packets and receiver_stats['received_packets'] > threshold
+        if result_queue.empty():
+            ret = False  # No frame is received.
         else:
-            ret = False
+            received_frames = result_queue.get()
+            if len(received_frames) != num_sent_frames:
+                ret = False
+            else:
+                # The radio driver may not append the FCF field to the received frame. Do not check the FCF field here.
+                ret = frame.tx_frame[:-4] == received_frames[0]['psdu'][:-4]
```

Comment:
I wonder what does the magical number `4` mean here?

Reply:
`4` is the length of the FCF field. I have changed it to a const value.